csrc/dd_arith.c

```diff
@@ -488,3 +494,244 @@ ddouble tanhq(ddouble a)
+    /* Strategy: Instead of using Taylor series to compute
+        arctan, we instead use Newton's iteration to solve
+        the equation
+
+            sin(z) = y/r    or    cos(z) = x/r
+
+        where r = sqrt(x^2 + y^2).
+        The iteration is given by
+
+            z' = z + (y - sin(z)) / cos(z)          (for equation 1)
+            z' = z - (x - cos(z)) / sin(z)          (for equation 2)
+
+        Here, x and y are normalized so that x^2 + y^2 = 1.
+        If |x| > |y|, then first iteration is used since the
+        denominator is larger.  Otherwise, the second is used.
+    */
+
+    if (iszeroq(x)) {
+
+        if (iszeroq(y)) {
+            /* Both x and y is zero. */
+            return Q_ZERO;
+    }
+
+        return (ispositiveq(y)) ? Q_PI_2 : negq(Q_PI_2);
+    } else if (iszeroq(y)) {
+        return (ispositiveq(x)) ? Q_ZERO : Q_PI;
+    }
+
+    if (equalqq(x, y)) {
+        return (ispositiveq(y)) ? Q_PI_4: negq(Q_3PI_4);
+    }
+
+    if (equalqq(x, negq(y))) {
+        return (ispositiveq(y)) ? Q_3PI_4 : negq(Q_PI_4);
+    }
+
+    ddouble r = sqrtq(addqq(sqrq(x), sqrq(y)));
+    ddouble xx = divqq(x, r);
+    ddouble yy = divqq(y, r);
+
+    /* Compute double precision approximation to atan. */
+    // ddouble z = (ddouble){atan2(y.hi, x.hi), 0.};
+    ddouble z = adddq(atan2(y.hi, x.hi), Q_ZERO);
+    ddouble sin_z, cos_z;
+
+    if (fabs(xx.hi) > fabs(yy.hi)) {
+        /* Use Newton iteration 1.  z' = z + (y - sin(z)) / cos(z)  */
+        sincos(z, &sin_z, &cos_z);
+        z = addqq(z, divqq(subqq(yy, sin_z), cos_z));
+    } else {
+
+        /* Use Newton iteration 2.  z' = z - (x - cos(z)) / sin(z)  */
+        sincos(z, &sin_z, &cos_z);
+        z = subqq(z, divqq(subqq(xx, cos_z), sin_z));
+    }
+
+  return z;
+}
+
+ddouble atan2dq(const double a, const ddouble b) {
+    return atan2qq(adddq(a, Q_ZERO), b);
+}
+
+ddouble atan2qd(const ddouble a, const double b) {
+    return atan2qq(a, adddq(b, Q_ZERO));
+}
+
+ddouble atanq(const ddouble a) {
+    return atan2qq(a, Q_ONE);
+}
+
+ddouble acosq(const ddouble a) {
+    ddouble abs_a = absq(a);
+
+    if (greaterqq(abs_a, Q_ONE)) {
+        return nanq();
+    }
+
+    if (isoneq(abs_a)) {
+        return (ispositiveq(a)) ? Q_ZERO : Q_PI;
+    }
+
+    return atan2qq(sqrtq(subdq(1.0, sqrq(a))), a);
+}
+
+ddouble asinq(const ddouble a) {
+    ddouble abs_a = absq(a);
+
+    if (greaterqd(abs_a, 1.0)) {
+        return nanq();
+    }
+
+    if (isoneq(abs_a)) {
+        return (ispositiveq(a)) ? Q_PI_2 : negq(Q_PI_2);
+    }
+
+    return atan2qq(a, sqrtq(subdq(1.0, sqrq(a))));
+}
+
+ddouble asinhq(const ddouble a) {
+    return logq(addqq(a,sqrtq(addqd(sqrq(a),1.0))));
+}
+
+ddouble acoshq(const ddouble a) {
+    if (lessqd(a, 1.0)) {
+        return nanq();
+    }
+
+    return logq(addqq(a, sqrtq(subqd(sqrq(a), 1.0))));
+}
+
+ddouble atanhq(const ddouble a) {
+    if (equalqd(a, -1.0)) {
+        return negq(infq());
+    } else if (equalqd(a, 1.0)) {
```

Comment:
Suggest use isone here
```suggestion
    } else if (isoneq(a)) {
```

Reply:
Fixed!